### PR TITLE
[bugfix] Resolve LearningRateMonitor + BackboneFinetuning

### LINKED ISSUE #7830: Error with `BackboneFinetuning` and `LearningRateMonitor` with `logging_interval="epoch"`
## 🐛 Bug

When using the backbone finetuning callback + learning rate moniter with logging interval = epoch an error is thrown after the backbone is unfrozen.

## Please reproduce using the BoringModel

```
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.callbacks.finetuning import BackboneFinetuning
from pytorch_lightning.callbacks.lr_monitor import LearningRateMonitor


class RandomDataset(Dataset):

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.backbone = torch.nn.Sequential(torch.nn.Linear(32, 32), torch.nn.ReLU(True))
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(self.backbone(x))

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        opt = torch.optim.Adam(self.layer.parameters(), lr=0.1)
        return [opt], [torch.optim.lr_scheduler.StepLR(opt, step_size=1)]


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        num_sanity_val_steps=0,
        max_epochs=6,
        weights_summary=None,
        callbacks=[BackboneFinetuning(unfreeze_backbone_at_epoch=4), LearningRateMonitor(logging_interval="epoch")]
    )
    trainer.fit(model, train_dataloader=train_data, val_dataloaders=val_data)
    trainer.test(model, test_dataloaders=test_data)


if __name__ == '__main__':
    run()
```

Error:
```
lr_monitor.py", line 160, in _extract_lr
    self.lrs[name].append(lr)
KeyError: 'lr-Adam/pg1'
Exception ignored in: <function tqdm.__del__ at 0x7ff5e83d7d30>
Traceback (most recent call last):
  File "/Users/ethan/miniconda3/lib/python3.8/site-packages/tqdm/std.py", line 1090, in __del__
  File "/Users/ethan/miniconda3/lib/python3.8/site-packages/tqdm/std.py", line 1303, in close
  File "/Users/ethan/miniconda3/lib/python3.8/site-packages/tqdm/std.py", line 1481, in display
  File "/Users/ethan/miniconda3/lib/python3.8/site-packages/tqdm/std.py", line 1093, in __repr__
  File "/Users/ethan/miniconda3/lib/python3.8/site-packages/tqdm/std.py", line 1443, in format_dict
TypeError: cannot unpack non-iterable NoneType object
```

### Expected behavior

No error

### Environment

PL master branch

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #7830

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/lr_monitor.py [hunk 1]
```diff
@@ -155,11 +155,26 @@ def _extract_stats(self, trainer, interval: str) -> Dict[str, float]:
 
         return latest_stat
 
-    def _extract_lr(self, param_group, name: str) -> Dict[str, float]:
+    def _extract_lr(self, trainer, param_group, name: str) -> Dict[str, float]:
         lr = param_group.get('lr')
-        self.lrs[name].append(lr)
+        try:
+            self.lrs[name].append(lr)
+        except KeyError:
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/lr_monitor.py:162]: it feels like `if name in self.lrs ... else ...` is better than try except but I could be wrong. maybe this is even slightly faster? ^^

File: tests/callbacks/test_lr_monitor.py [hunk 2]
```diff
@@ -278,3 +281,54 @@ def configure_optimizers(self):
...
+        def forward(self, x):
+            return self.layer(self.backbone(x))
+
+        def configure_optimizers(self):
+            opt = optim.Adam(self.layer.parameters(), lr=0.1)
+            opt_2 = optim.Adam(self.layer.parameters(), lr=0.1)
+            opt_3 = optim.Adam(self.layer.parameters(), lr=0.1)
+            return [opt, opt_2, opt_3
+                    ], [optim.lr_scheduler.StepLR(opt, step_size=1),
+                        optim.lr_scheduler.StepLR(opt_2, step_size=1)]
```
Comment (awaelchli) [inline tests/callbacks/test_lr_monitor.py:307]: formatting issue? looks broken 😄

**Review by awaelchli — APPROVED** [review]
LGTM just minor comments

File: pytorch_lightning/callbacks/lr_monitor.py [hunk 3]
```diff
@@ -168,35 +183,45 @@ def _extract_momentum(self, param_group, name: str, use_betas: bool) -> Dict[str
...
 
-                # Multiple schduler of the same type
-                while True:
-                    if name not in names:
-                        break
-                    i, name = i + 1, f'{opt_name}-{i}'
+            seen_optimizers.append(sch.optimizer)
+            optimizer_cls = type(sch.optimizer)
+            if scheduler['name'] is None:
+                if optimizer_cls not in seen_optimizer_types:
```
Comment (ethanwharris) [inline pytorch_lightning/callbacks/lr_monitor.py:206]: As `seen_optimizer_types` is a `defaultdict` you should just be able to do `seen_optimizer_types[optimizer_cls] += 1` and not need the if.

**Review by ethanwharris — APPROVED** [review]
LGTM, small comment

File: pytorch_lightning/callbacks/lr_monitor.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/lr_monitor.py:206]: Done !

File: pytorch_lightning/callbacks/lr_monitor.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/lr_monitor.py]: Thanks @awaelchli. Found a bug :)

File: tests/callbacks/test_lr_monitor.py (hunk 2 above)
Comment (tchaton) [reply tests/callbacks/test_lr_monitor.py]: Done !

File: pytorch_lightning/callbacks/lr_monitor.py [hunk 4]
```diff
@@ -146,20 +148,43 @@ def _extract_stats(self, trainer, interval: str) -> Dict[str, float]:
...
                     latest_stat.update(lr)
                     momentum = self._extract_momentum(
                         param_group=pg, name=f'{name}-momentum{suffix}', use_betas=use_betas
                     )
                     latest_stat.update(momentum)
 
+        print()
+        print(self.lrs)
+        print()
+
```
Comment (Borda) [inline pytorch_lightning/callbacks/lr_monitor.py:161]:
```suggestion
```

File: pytorch_lightning/callbacks/lr_monitor.py [hunk 5]
```diff
@@ -146,20 +148,43 @@ def _extract_stats(self, trainer, interval: str) -> Dict[str, float]:
...
         return latest_stat
 
-    def _extract_lr(self, param_group, name: str) -> Dict[str, float]:
+    def _extract_lr(self, trainer, param_group, name: str) -> Dict[str, float]:
         lr = param_group.get('lr')
-        self.lrs[name].append(lr)
+        print(trainer.current_epoch, name, lr)
```
Comment (Borda) [inline pytorch_lightning/callbacks/lr_monitor.py:166]:
```suggestion
```

File: tests/callbacks/test_lr_monitor.py [hunk 6]
```diff
@@ -278,3 +281,102 @@ def configure_optimizers(self):
...
+        callbacks=[TestFinetuning(), lr_monitor, Check()],
+        progress_bar_refresh_rate=0,
+        weights_summary=None,
+        checkpoint_callback=False
+    )
+    model = TestModel()
+    model.training_epoch_end = None
+    trainer.fit(model)
+
+    expected = [0.1, 0.09000000000000001, 0.08100000000000002, 0.07290000000000002, 0.06561000000000002]
```
Comment (Borda) [inline tests/callbacks/test_lr_monitor.py:369]: is there some rounding issue?

**Review by Borda — COMMENTED** [review]
is it ready? I see some remaining debugging prints etc.

File: tests/callbacks/test_lr_monitor.py (hunk 6 above)
Comment (tchaton) [reply tests/callbacks/test_lr_monitor.py]: Done !

File: pytorch_lightning/callbacks/lr_monitor.py [hunk 7]
```diff
@@ -155,11 +157,23 @@ def _extract_stats(self, trainer, interval: str) -> Dict[str, float]:
 
         return latest_stat
 
-    def _extract_lr(self, param_group, name: str) -> Dict[str, float]:
+    def _extract_lr(self, trainer, param_group, name: str) -> Dict[str, float]:
```
Comment (carmocca) [inline pytorch_lightning/callbacks/lr_monitor.py:160]: Unused trainer?

```suggestion
    def _extract_lr(self, param_group, name: str) -> Dict[str, float]:
```

File: pytorch_lightning/callbacks/lr_monitor.py [hunk 8]
```diff
@@ -155,11 +157,23 @@ def _extract_stats(self, trainer, interval: str) -> Dict[str, float]:
...
+        token = '/pg1'
+        for new_name in names:
+            if token in new_name:
+                old_n = new_name.replace(token, '')
+                if old_n in self.lrs:
+                    self.lrs[new_name] = self.lrs[old_n]
+                    del self.lrs[old_n]
+            else:
+                if new_name not in self.lrs:
+                    self.lrs[new_name] = []
```
Comment (carmocca) [inline pytorch_lightning/callbacks/lr_monitor.py:175]: Bit of refactoring, I believe it's the same

```suggestion
    def _remap_keys(self, names: List[str], token: str = '/pg1') -> None:
        for new in names:
            old = new.replace(token, '')
            if old in self.lrs:
                self.lrs[new] = self.lrs.pop(old)
            elif new == old:
                self.lrs[new] = []
```

File: pytorch_lightning/callbacks/lr_monitor.py [hunk 9]
```diff
@@ -168,35 +182,47 @@ def _extract_momentum(self, param_group, name: str, use_betas: bool) -> Dict[str
         self.last_momentum_values[name] = momentum
         return {name: momentum}
 
-    def _find_names(self, lr_schedulers) -> List[str]:
-        # Create uniqe names in the case we have multiple of the same learning
-        # rate schduler + multiple parameter groups
+    def _add_prefix(self, name, optimizer_cls, seen_optimizer_types) -> str:
```
Comment (carmocca) [inline pytorch_lightning/callbacks/lr_monitor.py:185]: types for the parameters?

File: pytorch_lightning/callbacks/lr_monitor.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/lr_monitor.py:207]: Doesn't look done 🧐

File: pytorch_lightning/callbacks/lr_monitor.py [hunk 10]
```diff
@@ -168,35 +182,47 @@ def _extract_momentum(self, param_group, name: str, use_betas: bool) -> Dict[str
...
 
-            # Multiple param groups for the same schduler
+            # Multiple param groups for the same scheduler
             param_groups = sch.optimizer.param_groups
 
             if len(param_groups) != 1:
-                for i, pg in enumerate(param_groups):
-                    temp = f'{name}/pg{i + 1}'
+                for i in range(len(param_groups)):
+                    temp = self._add_prefix(name, optimizer_cls, seen_optimizer_types)
```
Comment (carmocca) [inline pytorch_lightning/callbacks/lr_monitor.py:217]: Does this need to be inside the if/else and for-loop? Same for line 221

**Review by carmocca — CHANGES_REQUESTED** [review]

File: pytorch_lightning/callbacks/lr_monitor.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/lr_monitor.py:207]: Added back !

File: pytorch_lightning/callbacks/lr_monitor.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/lr_monitor.py:160]: Done !

File: pytorch_lightning/callbacks/lr_monitor.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/lr_monitor.py:175]: Done !

File: pytorch_lightning/callbacks/lr_monitor.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/lr_monitor.py:185]: Done !

File: pytorch_lightning/callbacks/lr_monitor.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/lr_monitor.py]: Done !

File: pytorch_lightning/callbacks/lr_monitor.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/lr_monitor.py]: Moved up.

File: pytorch_lightning/callbacks/lr_monitor.py [hunk 11]
```diff
@@ -146,7 +151,7 @@ def _extract_stats(self, trainer, interval: str) -> Dict[str, float]:
 
                 for i, pg in enumerate(param_groups):
                     suffix = f'/pg{i + 1}' if len(param_groups) > 1 else ''
-                    lr = self._extract_lr(param_group=pg, name=f'{name}{suffix}')
+                    lr = self._extract_lr(pg, name=f'{name}{suffix}')
```
Comment (carmocca) [inline pytorch_lightning/callbacks/lr_monitor.py:154]: nit: not a kwarg

```suggestion
                    lr = self._extract_lr(pg, f'{name}{suffix}')
```

File: tests/callbacks/test_lr_monitor.py [hunk 12]
```diff
@@ -278,3 +281,102 @@ def configure_optimizers(self):
...
+                assert num_param_groups == 3
+            elif trainer.current_epoch == 1:
+                assert num_param_groups == 4
+                assert list(lr_monitor.lrs.keys()) == ['lr-Adam-1', 'lr-Adam/pg1', 'lr-Adam/pg2']
+            elif trainer.current_epoch == 2:
+                assert num_param_groups == 5
+                assert list(lr_monitor.lrs.keys()) == ['lr-Adam/pg1', 'lr-Adam/pg2', 'lr-Adam-1/pg1', 'lr-Adam-1/pg2']
+            else:
+                expected = ['lr-Adam/pg1', 'lr-Adam/pg2', 'lr-Adam-1/pg1', 'lr-Adam-1/pg2', 'lr-Adam-1/pg3']
+                assert list(lr_monitor.lrs.keys()) == expected
```
Comment (carmocca) [inline tests/callbacks/test_lr_monitor.py:333]: nit: no need for keys

```suggestion
                assert list(lr_monitor.lrs) == ['lr-Adam-1', 'lr-Adam/pg1', 'lr-Adam/pg2']
            elif trainer.current_epoch == 2:
                assert num_param_groups == 5
                assert list(lr_monitor.lrs) == ['lr-Adam/pg1', 'lr-Adam/pg2', 'lr-Adam-1/pg1', 'lr-Adam-1/pg2']
            else:
                expected = ['lr-Adam/pg1', 'lr-Adam/pg2', 'lr-Adam-1/pg1', 'lr-Adam-1/pg2', 'lr-Adam-1/pg3']
                assert list(lr_monitor.lrs) == expected
```

**Review by carmocca — APPROVED** [review]
Minor comments

**Pull request merged**